### PR TITLE
Improve LigtningEnum, etc.

### LINKED ISSUE #12749: Improve `LightningEnum` and its deprecation
## Proposed refactor

- Use the built-in `Enum` capabilities to make `LightningEnum` simpler.
- Modify and rename the `_OnAccessEnumMeta` type to only call a `deprecate()` method (it's not used for anything else.
- Create a `DeprecatedEnum` base class and use that instead of constantly setting `metaclass=...` on deprecated enums.
- Correct attribution for `_OnAccessEnumMeta` (current attribution is to a website and article that falsely claims authorship of `Enum`).

### Motivation

Fixing the attribution (which came to my attention due to [an `aenum` issue](https://github.com/ethanfurman/aenum/issues/18), and efficient use of `Enum`.

cc @justusschock @awaelchli @rohitgr7 @tchaton

### PR DESCRIPTION
**Pull request by ethanfurman**
Fixes #12749 

- change `from_str` to use `__members__` for enum names
- change `from_str` to use built-in `__getitem__` syntax for names
- make `_OnAccessEnumMeta` more robust and efficient
- create `DeprecatedEnum` as a base class
- change `DistributedType` and `DeviceType` to use `DeprecatedEnum`

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/utilities/enums.py [hunk 1]
```diff
@@ -68,6 +68,12 @@ def __call__(cls, *args: Any, **kwargs: Any) -> Any:
         return obj
 
 
+class DeprecatedEnum(LightningEnum, metaclass=DeprecatedEnumMeta):
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/enums.py:71]:
```suggestion
class DeprecatedEnum(LightningEnum, metaclass=DeprecatedEnumMeta):
```
let's keep this protected?

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/utilities/enums.py (hunk 1 above)
Comment (ethanfurman) [reply pytorch_lightning/utilities/enums.py:71]: I don't see a change -- what is your concern and suggestion?

Comment (rohitgr7) [reply pytorch_lightning/utilities/enums.py:71]: updated!

Comment (ethanfurman) [reply pytorch_lightning/utilities/enums.py]: I also added an `_` to `DeprecatedEnumMeta`.

**Review by awaelchli — APPROVED** [review]
Someone is playing with metaclasses! <3

**Review by carmocca — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Pull request merged**